[clang-cl] Document behavior difference of strict aliasing in clang-cl vs clang.

huihzhang opened the pull request:
Add an user note to pass -fstrict-aliasing explicitly for clang-cl at optimization levels O3 and below to enable TBAA.
Also add help text for -f[no-]strict-aliasing.

huihzhang committed:
[clang-cl] Document behavior difference of strict aliasing in clang-cl vs clang.

Add an user note to pass -fstrict-aliasing explicitly for clang-cl at
optimization levels O3 and below to enable TBAA.
Also add help text for -f[no-]strict-aliasing.

zmodem commented on clang/include/clang/Driver/Options.td:
```diff
@@ -3152,7 +3152,8 @@ defm diagnostics_show_line_numbers : BoolFOption<"diagnostics-show-line-numbers"
 def fno_strict_aliasing : Flag<["-"], "fno-strict-aliasing">, Group<f_Group>,
-  Flags<[NoXarchOption]>, Visibility<[ClangOption, CLOption, DXCOption]>;
+  Flags<[NoXarchOption]>, Visibility<[ClangOption, CLOption, DXCOption]>,
+  HelpText<"Disable optimizations based on strict aliasing rules (default)">;
```
Since the defaultness depends on which driver mode is used (clang-cl or regular clang), we should probably leave out the `(default)` part.

zmodem commented on clang/docs/UsersManual.rst:
```diff
@@ -4722,3 +4724,17 @@ The Visual C++ Toolset has a slightly more elaborate mechanism for detection.
+^^^^^^^^^^^^^^^
+
+Strict aliasing (TBAA) is turned off by default in clang-cl at optimization
+levels O3 and below, and turned on at Ofast. Whereas in clang, strict aliasing
```
clang-cl uses MSVC's `-O` flag, so there is no `-O3` or `-Ofast`. So unless I'm missing something, I think we should say that it's always off by default in clang-cl.

zmodem reviewed (changes requested):
Thanks for the PR!

huihzhang committed:
[clang-cl] Document behavior difference of strict aliasing in clang-cl vs clang.

Add an user note to pass -fstrict-aliasing explicitly for clang-cl
to enable TBAA.
Also add help text for -f[no-]strict-aliasing.

huihzhang commented:
Fixed up documentation based on review comments.

zmodem reviewed (approved):
Looks good to me. Thanks!
